### PR TITLE
Feat bs monitor

### PR DESCRIPTION
**Pull request by constwz**
### Description

Add statistics for blocksyncer DB. A new metadata interface can obtain statistics

### Rationale

Help us identify problems

### CODE REVIEW

File: modular/blocksyncer/blocksyncer_options.go [hunk 1]
```diff
@@ -374,6 +396,52 @@ func (b *BlockSyncerModular) prepareMasterFlagTable() error {
 	return nil
 }
 
+func (b *BlockSyncerModular) dbStatistics(ctx context.Context) {
+	duration := b.DataStatisticsDuration
+	if duration == 0 {
+		duration = 3600
```
Comment (ruojunm) [inline modular/blocksyncer/blocksyncer_options.go:402]: it looks a bit frequent.  "count" in sql will hold a exclusive lock which prevent newly records from inserting.

File: modular/blocksyncer/database/db_statistics.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	result := make([]int64, 0, bsdb.ObjectsNumberOfShards)
+	step := int64(1000)
+	for i := 0; i < bsdb.ObjectsNumberOfShards; i++ {
+		sum := int64(0)
+		primaryKey := int64(0)
+		count := int64(0)
+		for {
+			var err error
+			tmpDB := db.Db.Table(bsdb.GetObjectsTableNameByShardNumber(i))
+			if needSeal {
```
Comment (ruojunm) [inline modular/blocksyncer/database/db_statistics.go:32]: the condition might miss some updating records.  
What if an old record (the id is small) is updated recently?

File: modular/blocksyncer/database/db_statistics.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import (
+	"context"
+	"time"
+
+	"github.com/bnb-chain/greenfield-storage-provider/pkg/log"
+
+	"github.com/forbole/juno/v4/models"
+	"gorm.io/gorm/clause"
+
+	"github.com/bnb-chain/greenfield-storage-provider/store/bsdb"
```
Comment (BarryTong65) [inline modular/blocksyncer/database/db_statistics.go:12]: should format the imports.

File: modular/blocksyncer/blocksyncer_options.go [hunk 4]
```diff
@@ -374,6 +396,52 @@ func (b *BlockSyncerModular) prepareMasterFlagTable() error {
...
+	var totalCount, sealCount []int64
+	var err error
+	for range ticker.C {
+		latestBlockHeightAny := Cast(b.parserCtx.Indexer).GetLatestBlockHeight().Load()
+		latestBlockHeight := latestBlockHeightAny.(int64)
+		if latestBlockHeight < 10 {
+			continue
+		}
+		if b.BlockResultStorage && latestBlockHeight-b.MaxBlockNum > 0 {
+			if delErr := db.Cast(b.parserCtx.Database).DeleteBlockResult(ctx, latestBlockHeight-b.MaxBlockNum); delErr != nil {
```
Comment (ruojunm) [inline modular/blocksyncer/blocksyncer_options.go:415]: The process of deleting stale block result data can be separated from block-syncer main thread by using a scheduler?

File: modular/blocksyncer/blocksyncer_options.go (hunk 1 above)
Comment (ruojunm) [reply modular/blocksyncer/blocksyncer_options.go:402]: ideally use read-only mysql instance

File: modular/blocksyncer/database/db_statistics.go (hunk 2 above)
Comment (ruojunm) [reply modular/blocksyncer/database/db_statistics.go:32]: 1. do we have index on update_at ?
2. should we include id range into the condition?

File: modular/gater/metadata_handler.go [hunk 5]
```diff
@@ -2781,3 +2782,46 @@ func (g *GateModular) getBucketSizeHandler(w http.ResponseWriter, r *http.Reques
...
+		} else {
+			metrics.ReqCounter.WithLabelValues(GatewayTotalSuccess).Inc()
+			metrics.ReqTime.WithLabelValues(GatewayTotalSuccess).Observe(time.Since(startTime).Seconds())
+		}
+	}()
+
+	ctx := context.Background()
+	queryParams := r.URL.Query()
+	blockHeightStr := queryParams.Get("block_height")
+	blockHeight, err = strconv.ParseInt(blockHeightStr, 10, 64)
```
Comment (BarryTong65) [inline modular/gater/metadata_handler.go:2808]: recommend to use util.StringToInt64 to parse the int64 for better consistency
```
	startTimestampUs, err = util.StringToInt64(vars["start_ts"])
	if err != nil {
		log.CtxErrorw(ctx, "failed to parse start_ts query", "error", err)
		err = ErrInvalidQuery
		return
	}
```

File: modular/gater/router.go [hunk 6]
```diff
@@ -262,6 +263,9 @@ func (g *GateModular) RegisterHandler(router *mux.Router) {
 	// List Bucket Read Quota Count
 	router.Path("/").Name(getBucketReadQuotaCountRouterName).Methods(http.MethodGet).Queries(ListBucketReadCountQuery, "").HandlerFunc(g.getBucketReadQuotaCountHandler)
 
+	// Get BsDB data statistics Info
+	router.Path("/").Name(getBsDBDataInfo).Methods(http.MethodGet).Queries(BsDBInfo, "").HandlerFunc(g.getBsDBDataInfoHandler)
```
Comment (BarryTong65) [inline modular/gater/router.go:267]: To improve naming consistency, consider changing `BsDBInfo` to `BsDBInfoQuery`, similar to `ListBucketReadCountQuery` and `GetPaymentByBucketIDQuery`, to clearly indicate its use for routing.

File: modular/gater/router_test.go [hunk 7]
```diff
@@ -749,6 +749,14 @@ func TestRouters(t *testing.T) {
 			shouldMatch:      true,
 			wantedRouterName: delegateCreateFolderRouterName,
 		},
+		{
+			name:             "get block syncer data info",
+			router:           gwRouter,
+			method:           http.MethodGet,
+			url:              fmt.Sprintf("%s%s/?%s&%s", scheme, testDomain, BsDBInfo, "block_height"),
```
Comment (BarryTong65) [inline modular/gater/router_test.go:756]: ditto

File: modular/gater/const.go [hunk 8]
```diff
@@ -260,6 +260,7 @@ const (
 	Delegate = "delegate"
 	// CreateFolderQuery defines the delegate create folder path
 	CreateFolderQuery = "create-folder"
+	BsDBInfo          = "bsdb-info"
```
Comment (BarryTong65) [inline modular/gater/const.go:263]: ditto

**Review by ruojunm — APPROVED** [review]

**Review by BarryTong65 — APPROVED** [review]

**Pull request merged**